ide/lsp.client/src/org/netbeans/modules/lsp/client/bindings/FoldManagerImpl.java

```diff
@@ -113,49 +113,56 @@ public void run(LSPBindings bindings, FileObject file) {
         EditorCookie ec = file.getLookup().lookup(EditorCookie.class);
         Document doc = ec != null ? ec.getDocument() : null;
         if (doc == null) {
-            return ;
+            return;
         }
-        Set<FoldingRangeInfo2> foldingRangesSeen = new HashSet<>();
+
         List<FoldingRange> ranges = computeRanges(bindings, file);
+        List<FoldInfo> infos = computeInfos(doc, ranges);
+        SwingUtilities.invokeLater(() -> {
+            doc.render(() -> {
+                operation.getHierarchy().render(() -> {
+                    try {
+                        operation.update(infos, null, null);
+                    } catch (BadLocationException ex) {
+                        LOG.log(Level.FINE, null, ex);
+                    }
+                });
+            });
+        });
+    }
+    
+    static List<FoldInfo> computeInfos(Document doc, List<FoldingRange> ranges) {
+        Set<FoldingRangeInfo2> foldingRangesSeen = new HashSet<>();
         List<FoldInfo> infos = new ArrayList<>();
         if (ranges != null) {
             for (FoldingRange r : ranges) {
                 int start = Utils.getOffset(doc, new Position(r.getStartLine(), r.getStartCharacter() != null ? r.getStartCharacter() : 0));
```

Comment:
This is probably the place @matthiasblaesing  noted in the earlier comment: if `startCharacter` is nto defined, then, according to the documentation, it should equal to the length of the start line == the end of that line.

Please either correct the expression, or make a comment that the code diverges from the spec and why exactly is so.